src/lib/planviz/abstract_visualizer.hpp

```diff
@@ -0,0 +1,156 @@
+  std::string shape = GraphvizShape::Rectangle;
+  uint8_t pen_width = 1u;
+};
+
+struct VizEdgeInfo {
+  std::string label;
+  std::string color = GraphvizColor::White;
+  std::string font_color = GraphvizColor::White;
+  uint8_t pen_width = 1u;
+};
+
+template <typename GraphBase>
+class AbstractVisualizer {
+  //                                  Edge list    Vertex list   Directed graph
+  using Graph = boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS,
+                                      // Vertex info Edge info    Graph info
+                                      VizVertexInfo, VizEdgeInfo, VizGraphInfo>;
+
+ public:
+  AbstractVisualizer() : AbstractVisualizer(GraphvizConfig{}, VizGraphInfo{}, VizVertexInfo{}, VizEdgeInfo{}) {}
+
+  AbstractVisualizer(GraphvizConfig graphviz_config, VizGraphInfo graph_info, VizVertexInfo vertex_info,
+                     VizEdgeInfo edge_info)
+      : _graphviz_config(std::move(graphviz_config)),
+        _graph_info(std::move(graph_info)),
+        _default_vertex(std::move(vertex_info)),
+        _default_edge(std::move(edge_info)) {
+    // Add global Graph properties
+    _add_graph_property("rankdir", _graph_info.rankdir);
+    _add_graph_property("bgcolor", _graph_info.bg_color);
+    _add_graph_property("ratio", _graph_info.ratio);
+
+    // Add vertex properties
+    _add_property("node_id", boost::vertex_index);
+    _add_property("color", &VizVertexInfo::color);
+    _add_property("label", &VizVertexInfo::label);
+    _add_property("shape", &VizVertexInfo::shape);
+    _add_property("fontcolor", &VizVertexInfo::font_color);
+    _add_property("penwidth", &VizVertexInfo::pen_width);
+
+    // Add edge properties
+    _add_property("color", &VizEdgeInfo::color);
+    _add_property("fontcolor", &VizEdgeInfo::font_color);
+    _add_property("label", &VizEdgeInfo::label);
+    _add_property("penwidth", &VizEdgeInfo::pen_width);
+  }
+
+  void visualize(const GraphBase& graph_base, const std::string& graph_filename, const std::string& img_filename) {
+    _build_graph(graph_base);
+    std::ofstream file(graph_filename);
+    boost::write_graphviz_dp(file, _graph, _properties);
+
+    auto executable = _graphviz_config.layout;
+    auto format = _graphviz_config.format;
+
+    auto cmd = executable + " -T" + format + " " + graph_filename + " > " + img_filename;
+    auto ret = system(cmd.c_str());
+
+    Assert(ret == 0, "Calling graphviz' " + executable +
+                         " failed. Have you installed graphviz "
+                         "(apt-get install graphviz / brew install graphviz)?");
+    // We do not want to make graphviz a requirement for Hyrise as visualization is just a gimmick
+  }
+
+ protected:
+  virtual void _build_graph(const GraphBase& graph_base) = 0;
+
+  template <typename T>
+  void _add_vertex(const T& vertex) {
+    _add_vertex(vertex, _default_vertex);
+  }
+
+  template <typename T>
+  void _add_vertex(const T& vertex, const std::string& label) {
```

Comment:
Can we have a default parameter for `label` so that we can get rid of the one-parameter version above?

Reply:
sure